### PR TITLE
New <Grid> implementation

### PR DESCRIPTION
**Pull request by carldebilly**
Fixes https://github.com/unoplatform/uno/issues/432
Fixes https://github.com/unoplatform/uno/issues/3483
Fixes https://github.com/unoplatform/uno/issues/4727
Fixes https://github.com/unoplatform/uno/issues/4919
Fixes https://github.com/unoplatform/uno/issues/4212
Fixes https://github.com/unoplatform/uno/issues/1885
May fix https://github.com/unoplatform/uno/issues/2201 (unchecked yet)
May fix https://github.com/unoplatform/uno/issues/2088 (unchecked yet)
May fix https://github.com/unoplatform/uno/issues/1611 (unchecked yet)
May fix https://github.com/unoplatform/uno/issues/5630 (unchecked yet - no repro in issue)
Should fix https://github.com/unoplatform/nventive-private/issues/164 (need to test in concerned app)
Related to https://github.com/unoplatform/uno/issues/4231
Related to https://github.com/unoplatform/uno/issues/2603

# Feature
Old implementation of `<Grid />` has been scrapped in the PR and replaced by the code from UWP/WinUI,

A lot of bugs and performance improvements should be fixed by this.

## PR Checklist

Please check if your PR fulfills the following requirements:

- ~~[ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)~~
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [X] Contains **NO** breaking changes
- [X] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [X] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/Grid/ColumnDefinition.cs [hunk 1]
```diff
@@ -60,12 +64,75 @@ public double ActualWidth
 		{
 			get
 			{
-				var parent = this.GetParent();
-				var result = (parent as Grid)?.GetActualWidth(this) ?? 0d;
-				return result;
+				//var parent = this.GetParent();
+				//var result = (parent as Grid)?.GetActualWidth(this) ?? 0d;
+				//return result;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/ColumnDefinition.cs:69]: Remove ?

File: src/Uno.UI/UI/Xaml/Controls/Grid/DefinitionCollectionBase.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+using System.Collections.Generic;
+
+namespace Windows.UI.Xaml.Controls
+{
+	internal interface DefinitionCollectionBase : IList<DefinitionBase>
+	{
+		DefinitionBase GetItem(int index);
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/DefinitionCollectionBase.cs:7]: Should be internal as well ?
```suggestion
		internal DefinitionBase GetItem(int index);
```

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.Properties.cs [hunk 3]
```diff
@@ -4,10 +4,11 @@
 using System.Collections.Generic;
 using System.Text;
 using Uno.UI;
+using Uno.UI.Extensions;
 using Uno.UI.Xaml;
 
 #if XAMARIN_ANDROID
-using View = Android.Views.View;
+using View = Windows.UI.Xaml.UIElement;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.Properties.cs:11]: This will be a breaking change. Why is it needed ?

File: src/SamplesApp/SamplesApp.UnitTests.Shared/Controls/UITests/Views/Controls/SampleChooserControl.xaml [hunk 4]
```diff
@@ -340,7 +340,7 @@
 
 				<!--NAVIGATION BUTTONS-->
 				<Border Background="{StaticResource Color08Brush}"
-						Grid.ColumnSpan="2"
```
Comment (jeromelaban) [inline src/SamplesApp/SamplesApp.UnitTests.Shared/Controls/UITests/Views/Controls/SampleChooserControl.xaml:343]: Why is this needed ?

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs [hunk 5]
```diff
@@ -1,351 +1,1805 @@
...
-				var width = ColumnDefinitions[i].Width;
-				if (width.IsAuto||width.IsStar)
+				//empty collection defaults to single row
+				//CValue value;
+				//value.WrapObjectNoRef(null);
+				//CREATEPARAMETERS param(this.
+				//GetContext(), value);
+				//ColumnDefinitionCollection.Create((CDependencyObject)(m_pColumns), &param);
+				//ColumnDefinition.Create((CDependencyObject)(emptyColumn), &param);
+				//m_pColumns.Append(emptyColumn);
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:184]: Remove ?

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs [hunk 6]
```diff
@@ -1,351 +1,1805 @@
...
+			else
+			{
+				m_pColumns = m_pColumnDefinitions;
+				//m_pColumnDefinitions.AddRef();
+			}
+
+			// Cleanup
+			//ReleaseInterface(emptyRow);
+			//ReleaseInterface(emptyColumn);
+			// return hr;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:198]: Remove

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs [hunk 7]
```diff
@@ -1,351 +1,1805 @@
...
+				//ColumnDefinition.Create((CDependencyObject)(emptyColumn), &param);
+				//m_pColumns.Append(emptyColumn);
+				m_pColumns = new ColumnDefinitionCollection();
+				emptyColumn = new ColumnDefinition();
+				m_pColumns.Add(emptyColumn);
+			}
+			else
+			{
+				m_pColumns = m_pColumnDefinitions;
+				//m_pColumnDefinitions.AddRef();
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:192]: Remove

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs [hunk 8]
```diff
@@ -1,351 +1,1805 @@
...
 		{
-			if (requester == null)
+
+			SpanStoreStackVector spanStore = new SpanStoreStackVector();
+
+			if (cellsHead >= cellCount)
 			{
-				// The Grid itself called invalidate
-				return this.IsHeightConstrainedSimple();
+				//goto Cleanup;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:374]: Remove

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs [hunk 9]
```diff
@@ -1,351 +1,1805 @@
...
 				}
-			}
-			for (int i = row; i < row + rowSpan; i++)
+
+				cellsHead = cellCacheVector[cellsHead].m_next;
+
+			} while (cellsHead < cellCount);
+
+			//Go through the spanned rows/columns allocating sizes.
+			//for (auto & entry : spanStore.m_vector)
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:433]: Probably should remove all those as well

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs [hunk 10]
```diff
@@ -1,351 +1,1805 @@
...
+			// Locking the row and columns definitions to prevent changes by user code
+			// during the arrange pass.
+			LockDefinitions();
+			//var scopeGuard = wil.scope_exit([&]
+			//{
+			//	delete[] m_ppTempDefinitions;
+			//	m_ppTempDefinitions = null;
+			//	m_cTempDefinitions = 0;
+			//	UnlockDefinitions();
+			//});
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:1341]:
```suggestion
```

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs [hunk 11]
```diff
@@ -1,351 +1,1805 @@
...
 		}
+
+		//void OnPropertyChanged(PropertyChangedParams args)
+		//{
+		//	//Panel.OnPropertyChanged(args);
+
+		//	if (args.Property == Grid.RowDefinitionsProperty
+		//	    || args.Property == Grid.ColumnDefinitionsProperty)
+		//	{
+		//		InvalidateDefinitions();
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:1725]: Implemented somewhere else?

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.h.cs [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+
+		//class CColumnDefinitionCollection;
+
+		//class CDefinitionBase;
+
+		//class CDefinitionCollectionBase;
+
+		//class CDOCollection;
+
+		//class CRowDefinitionCollection;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.h.cs:32]: Remove

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.h.cs [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+		private enum GridFlags : byte
+		{
+			None = 0x00,
+			HasStarRows = 0x01,
+			HasStarColumns = 0x02,
+			HasAutoRowsAndStarColumn = 0x04,
+			DefinitionsChanged = 0x08,
+		}
+
+		//DEFINE_ENUM_FLAG_OPERATORS(GridFlags);
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.h.cs:44]: Remove

File: src/Uno.UI/UI/Xaml/Shapes/Shape.layout.cs [hunk 14]
```diff
@@ -640,7 +640,8 @@ private static (float x, float y) ComputeScaleFactors(Size geometrySize, ref Siz
 			}
 			else
 			{
-				x = (float)((renderSize.Width - strokeThickness) / geometrySize.Width);
+				var renderWidthWithStrokeThickness = Math.Max(renderSize.Width - strokeThickness, 0);
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Shapes/Shape.layout.cs:643]: unrelated ?

**Comment by robloo** [comment]
Awesome to see this coming together! I always wondered why the WPF code wasn't pulled in a few years ago but since you have access to WinUI source all the better!

File: src/SamplesApp/SamplesApp.UnitTests.Shared/Controls/UITests/Views/Controls/SampleChooserControl.xaml (hunk 4 above)
Comment (carldebilly) [reply src/SamplesApp/SamplesApp.UnitTests.Shared/Controls/UITests/Views/Controls/SampleChooserControl.xaml:343]: To be able to test at some point I had to prevent any overflow. In this case, there's only one column, so the ColumnSpan were crashing my first version of the Grid. That's fixed now. Do you think I should revert this, since it's pratically a non-sense?

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs (hunk 5 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:184]:
```suggestion
```

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs (hunk 6 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:198]:
```suggestion
```

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs (hunk 7 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:192]:
```suggestion
```

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs (hunk 8 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:374]:
```suggestion
```

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs (hunk 9 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:433]:
```suggestion
```

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.h.cs (hunk 12 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Grid/Grid.h.cs:32]:
```suggestion
```

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.h.cs (hunk 13 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Grid/Grid.h.cs:44]:
```suggestion
```

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs (hunk 11 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Grid/Grid.cs:1704]: Yep.  It's in the `OnGenericPropertyChanged()` of the `Grid.Properties.cs` file.

File: src/Uno.UI/UI/Xaml/Controls/Grid/ColumnDefinition.cs (hunk 1 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Controls/Grid/ColumnDefinition.cs:73]:
```suggestion
```

**Review by MartinZikmund — APPROVED** [review]
YesNiceGIF

**Comment by robloo** [comment]
This is great: several longstanding issues are now closed and views carried over from UWP that would previously crash the app now load just fine. I was hoping a new grid implementation would save a lot of debugging/bug fixes and it has!

**Comment by francoistanguay** [comment]
> This is great: several longstanding issues are now closed and views carried over from UWP that would previously crash the app now load just fine. I was hoping a new grid implementation would save a lot of debugging/bug fixes and it has!

Thank you, this great feedback and always good to hear our team's efforts are appreciated and leveraged!
